### PR TITLE
Make all functions and variables defined/declared globally

### PR DESCRIPTION
**Pull request by muzimuzhi**
This PR changes all locally defined and updated sensible l3kernel functions and variables, both public and internal, to be defined or declared globally. 

Commit https://github.com/latex3/latex3/commit/6f301f13dadb7585b43c1cb744fc232a0a36fdd9 adds a plainTeX test that loads `expl3-generic.tex` in a group with `\tracingrestores` turned on. I think it dumps too many lines hence am considering dropping it in final state.

TODO:
- Check whether some functions in `l3basic.dtx` may be defined by `\cs_new_...`, rather than `\cs_gset`.
- Can the deprecated `\keys_set_filter:...` be defined by `\cs_new_...` too?

Remaining local definitions, which I think are expected to be local-only:
```
$ rg -F '=undefined}' -g '*/m3plain002.tlg'
testfiles-plain/m3plain002.tlg
163:{restoring \__text_loop:Nn=undefined}
194:{restoring \@protected@testopt=undefined}
390:{restoring \cite =undefined}
1319:{restoring \@unexpandable@protect=undefined}
1600:{restoring \__skip_tmp:w=undefined}
1678:{restoring \__file_seen_/dev/null:=undefined}
1679:{restoring \__file_name=/dev/null=undefined}
1680:{restoring \__file_name=nul:=undefined}
1787:{restoring \__iow_tmp:w=undefined}
1862:{restoring \__msg_tmp:w=undefined}
3268:{restoring \__tl_tmp:w=undefined}
3393:{retaining \__tl_tmp:w=undefined}
3712:{restoring \__quark_tmp:w=undefined}
3843:{restoring \ExplSyntaxOn=undefined}
3850:{restoring \ExplSyntaxOff=undefined}
3852:{restoring \ver@expl3-code.tex=undefined}
3853:{restoring \AtBeginDocument=undefined}
3855:{restoring \l@expl@tidy@tl=undefined}
3856:{restoring \ver@expl3-generic.tex=undefined}
3857:{restoring \ExplLoaderFileDate=undefined}
3858:{restoring \ExplFileDate=undefined}
```

### CODE REVIEW

**Comment by muzimuzhi** [comment]
> ```
> 3843:{restoring \ExplSyntaxOn=undefined}
> 3850:{restoring \ExplSyntaxOff=undefined}
> ```

Do we want these to be defined globally at first?

File: l3kernel/testfiles-plain/m3plain002.tlg [hunk 1]
```diff
@@ -0,0 +1,3858 @@
+This is a generated file for the LaTeX (2e + expl3) validation system.
+Don't change this file in any respect.
+{restoring \TYPE=\long macro:#1->\begingroup \long \def \TYPE \ETC.}
+{restoring \l__kernel_expl_bool=\char"1}
+{retaining \__opacity_backend_fill_stroke:ee=\protected\long macro:->\exp_args:Nee \__opacity_backend_fill_stroke:nn }
+{retaining \__pdf_backend_object_now:ne=\protected\long macro:->\exp_args:Nne \__pdf_backend_object_now:nn }
+{retaining \__pdf_backend_object_write:nne=\protected\long macro:->\exp_args:Nnne \__pdf_backend_object_write:nnn }
+{restoring \l_graphics_search_ext_seq=undefined}
```
Comment (muzimuzhi) [inline l3kernel/testfiles-plain/m3plain002.tlg:8]: The treatment for `\l_graphics_search_ext_seq` is split into #1469. This makes the m3plain002 output sensitive to whether #1469 is merged.

**Comment by muzimuzhi** [comment]
> Commit [6f301f1](https://github.com/latex3/latex3/commit/6f301f13dadb7585b43c1cb744fc232a0a36fdd9) adds a plainTeX test that loads `expl3-generic.tex` in a group with `\tracingrestores` turned on. I think it dumps too many lines hence am considering dropping it in final state.

Is there a way to track lines containing `=undefined}` only? If so, I'm willing to keep `m3plain002.lvt`.

It seems this calls for a new primitive `\tracingstream`, similar to `\showstream` added in TeX Live 2022.

**Comment by josephwright** [comment]
I've rebased locally to fix the conflict and pushed: all done

**Comment by muzimuzhi** [comment]
Remaining question: Do we want `\ExplSyntaxOn` and `\ExplSyntaxOff` globally defined at first time?

https://github.com/latex3/latex3/blob/59d17d1932255ab1fe9359d5799cccb7a6306953/l3kernel/l3bootstrap.dtx#L317
https://github.com/latex3/latex3/blob/59d17d1932255ab1fe9359d5799cccb7a6306953/l3kernel/l3bootstrap.dtx#L272

**Comment by josephwright** [comment]
@muzimuzhi Looking at the diffs I've had since merging this, I wonder if we might be better simply erroring if not loaded at top-level? Whilst one can arrange that all functions are globally-defined, variables with local scope semantics have their _value_ reset. That's likely to lead to errors/issues if those parts of the code are then used outside of the loading group. Thoughts?

**Comment by muzimuzhi** [comment]
> I wonder if we might be better simply erroring if not loaded at top-level?

Agreed. `m3plain002.lvt` was added mainly for catching remaining local declarations and I too don't think loading l3kernel inside group(s) is ever officially supported.

**Comment by FrankMittelbach** [comment]
I would error and not attempting to carry on.

**Pull request closed without merging**